lib/spack/spack/solver/asp.py

```diff
@@ -2163,6 +2159,55 @@ def literal_specs(self, specs):
             self.gen.fact(fn.concretize_everything())
 
 
+def _get_versioned_specs_from_pkg_requirements():
+    """If package requirements mention versions that are not mentioned
+    elsewhere, then we need to collect those to mark them as possible
+    versions.
+    """
```

Comment:
@tgamblin @scheibelp Sorry if I came late to reviewing this, and previously stopped at reading the title. My understanding was that this PR allowed unknown git versions to be used in package requirements, which is :+1: . 

Reading the implementation (while resolving conflicts in other PRs) it seems to me that this PR makes any non-existing version in package requirements an allowed one. I think this is a bug. For example:
```yaml
packages:
  zlib:
    require: "@2.0" # This version doesn't exist and might be the result of a typo
```
With the previous behavior on develop we would get:
```console
$ spack spec zlib
==> Error: Cannot satisfy the requirements in packages.yaml for the 'zlib' package. You may want to delete them to proceed with concretization. To check where the requirements are defined run 'spack config blame packages'
```
which, despite the error message that can be improved, is what I expect by a requirement. This PR made it such that:
```console
$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@2.0%gcc@9.4.0+optimize+pic+shared build_system=makefile arch=linux-ubuntu20.04-icelake
```
which in my opinion is wrong, since on version 2.0 exists for `zlib`. Was that on purpose?

Reply:
> # This version doesn't exist and might be the result of a typo

> which in my opinion is wrong, since no version 2.0 exists for zlib

If the user says it must be `zlib@2.0` then I think that's what it should be: if you say `spack spec zlib@2.0` on the command line then that's what we get (and Spack tries to find 2.0 from zlib's download link).

If nothing else, right now if a user says `require: foo@hash=1.5` and `1.5` doesn't exist in the package.py, then we want Spack to still allow it. I think it's consistent to extend that to all versions mandated in the config.

Printing a warning message might be in order if the package can't be fetched, but otherwise I would say this is the desired behavior. If you want to prevent declaration of number versions that don't exist (e.g. when not associated with a git hash), then that would work for the use cases I created it for, but would be inconsistent (for the reasons outlined above).

Reply:
I think it's inconsistent somehow. If the user says `+shared` we don't allow a `shared` variant where it doesn't exist. I think the only case where popping a non-git version into existence make sense are externals, since it's software already installed - and somebody put the information on its version in `packages.yaml`.

Reply:
If I do

```
$ spack spec zlib@2.0
Input spec
--------------------------------
zlib@2.0

Concretized
--------------------------------
zlib@2.0%apple-clang@12.0.0+optimize+pic+shared build_system=makefile arch=darwin-bigsur-skylake
```

do make sure, are you also arguing for not doing that anymore?

Reply:
The problem is ultimately the ambiguity of `@2.0` syntax. I can sort of get Peter's interpretation, but I must say I find it a bit far-fetched to turn a requirement into a version definition. At least to me it looks different from specifying a version on the command line.

In any case, with #36273 the ambiguity can be resolved since it introduces syntax to distinguish between a version range and a specific version.

Reply:
> are you also arguing for not doing that anymore?

No, or at least not necessarily My points:
- Command line might be used for quick tests, so we should allow something like this (maybe gated to an argument, but that would be a separate discussion)
- We moved an early error to a later stage, where it may potentially hang waiting for an answer
- To require a given version and be sure Spack doesn't invent a new one, I am now forced to use a range. This is a change in behavior on how the configuration is treated.

That made me think this was a bug that slipped through. Do you agree that a configuration that previously would have failed almost immediately, can now potentially hang in CI?

Reply:
I agree with @alalazo it's much more likely the user makes a typo than that they intentionally request a non-existing version (and happen to know of this curious behavior of Spack). Personally I never use non-existing versions, and it's not _that_ hard to just "register" the version if you need something not there yet.

Regarding ambiguity I mean: you an interpret it as "use some 2.0.something" or "use 2.0 exactly". If you're reading it as "use 2.0.something" you would expect an error if there is no 2.0.something. If you read it as 2.0 exactly you would _maybe_ expect spack to make that version up (but apparently both @alalazo and me did not expect that to happen).

Reply:
> Personally I never use non-existing versions, and it's not that hard to just "register" the version if you need something not there yet.

Firstly by "register" I think you mean "add the version to package.py". Enough users that I interact with need this (i.e. the ability to request versions *not* in the package.py) enough that I consider it necessary; although for the users I'm thinking of, they just need `require: @hash=number` to do this, rather than `require: @number`.

I could see the argument that it's unlikely for users to want a numbered version not in the package.py, but like I said we already (i.e. before this PR) allowed Spack to do this in other circumstances (the command line, or when specified as a preference). 

Are you (@haampie and @alalazo) proposing:

* Forbid `spack spec zlib@2.0` (currently allowed)
* Forbid `packages:zlib:version:[2.0]` (currently allowed)
* Allow `spack spec zlib@hash=2.0` (pretending zlib has a git repo)
* Allow `packages:zlib:verssion:[hassh=2.0]`

?

Reply:
I would be fine already with:
- Allow `@<any git version>` in requirements to define the version
- Get back the previous behavior for non-git versions

and discuss the other points later. Right now any requirement on version is somehow self-fulfilling (and bound to fail later). 

---

That said, I think defining unknown versions make sense only for `git versions` or already installed software (buildcache, local DB, or externals). We can probably allow:
```
packages:zlib:version:[2.0]
```
But that shouldn't be defining a version 2.0 if it doesn't exist somewhere else. So, clingo would prefer to use 2.0 but if it isn't defined somewhere else, would use the next best.

Reply:
When we define an environment, we also validate that the concretization output matches all of the versions we specify. So that should, in theory, catch the fact that the defined default in packages doesn't match what we get. 

A warning would be nice though, that way we see the discrepancy faster. 

Also, having packages define default versions that are undefined in the package causes concretization errors. The version doesn't end up in the concretization facts, version checks in `when` clauses all fail for that version.

Reply:
@alalazo

> But that shouldn't be defining a version 2.0 if it doesn't exist somewhere else.

To be clear, that was the behavior in Spack prior to this PR: putting versions in the preferences defines them as possible versions (i.e. if `2.0` is a version preference, `spack spec zlib` will give you `zlib@2.0`). Not to say you can't undo that, but it will take (slightly more) than reverting parts of this PR to get what you want.

Therefore

> Get back the previous behavior for non-git versions

the "previous behavior" doesn't seem to be a valid given your own stated desires. I suggest outlining the things you want to be true (e.g. as done in https://github.com/spack/spack/pull/36347#discussion_r1153902072) and then we can make them true (ignoring however they were behaving in the past).

Reply:
> To be clear, that was the behavior in Spack prior to this PR [ .. ]

I know, sorry if the reply was confusing but:
1. My immediate worry is if we can get back the old behavior for non-git versions in requirements
2. Then you asked what I thought about the other parts (preferences, CLI) so I gave an answer to that too - but changes there can definitely wait

Also, see this issue we got today https://github.com/spack/spack/issues/36574 which is related to the discussion.

Reply:
It is related but FWIW if you check out 9a93f223d6 (the commit before this was merged) you can still reproduce https://github.com/spack/spack/issues/36574. In some sense, my goal will be to extend this to do the opposite of what was decided there (for `hash=number` versions at least).

Preventing users from introducing new non-hash versions in preferences will not conflict with this PR or the users I know of that would benefit from it, so I only oppose it on account of it being inconsistent (I am willing to be overridden on that).

Reply:
> It is related 

I said this because I assumed that package requirements would have the same problem, but they don't: https://github.com/spack/spack/pull/36589; therefore that issue is not related to this PR (or at least not caused by it).